horovod/_keras/elastic.py

```diff
@@ -54,6 +60,10 @@ def __init__(self, backend, state, *args):
         super(UpdateEpochStateCallbackImpl, self).__init__(*args)
         self.backend = backend
         self.state = state
+        self.initial_epoch = self.state.epoch
+
+    def on_train_begin(self, logs=None):
+        self.initial_epoch = self.state.epoch
 
     def on_epoch_end(self, epoch, logs=None):
-        self.state.epoch = epoch
+        self.state.epoch = self.initial_epoch + epoch + 1
```

Comment:
is `epoch` the epoch that ended? why `self.state.epoch = self.initial_epoch + epoch + 1`

Reply:
Yes, `epoch` is the epoch that just ended.  The idea is that at the end of the epoch, we want to `commit` to save our progress.  We could avoid having to do the `+ 1` here if we called this `on_epoch_begin`, but if we commit before incrementing the epoch number (but after resetting the batch number), then we would end up repeating the last epoch.  So this way we ensure that when we commit, the state is fully up to date in case an error occurs.

The `initial_epoch` is because the `epoch` provided by the callback is a relative epoch (we cannot offset it at an initial epoch).  So if we complete some number of epochs and then reset, we do not want to lose our progress.

I will add comments to explain this in the code.